Set up Node.js server with patient greeting CDS service enabled. Configure CORS on the application middleware and routes for the discovery endpoint and CDS Services

zplata opened the pull request:
Currently, the default CDS Services configured for the Sandbox are contained within the Sandbox repository. The services are hosted externally on `http://hooks.fhir.me` but we should start to extract these out of the Sandbox repository and into this repository. This allows for separating maintenance of the Sandbox and the default CDS Services. We can also set configuration in this project that allows for deployment to Google App Engine. 

This task includes setting up the Node server with the Express framework, defining CORS and the routes for the discovery and CDS Service endpoints, with implementation for the patient greeting service. Additionally, this task includes an `app.yaml` file which specifies configuration for deployment to Google App Engine for hosting.

kpshek commented on index.js:
```diff
@@ -0,0 +1,73 @@
+    response.status(400);
+  }
+
+  const cdsService = services[serviceEndpoint];
```
In either of the above error scenarios (404 or 500), we should be returning immediately. Instead, execution flow continues to this `services` map lookup.

For instance, run the server locally and do a POST to http://localhost:3000/cds-services/foo. You'll see a log error of:

```
TypeError: Cannot read property 'serviceResponse' of undefined
    at app.post (/Users/shek/dev/sandbox-cds-services/index.js:68:13)
```

kpshek commented on index.js:
```diff
@@ -0,0 +1,73 @@
+
+app.use((request, response, next) => {
+  if (request.method !== 'GET' && request.headers && request.headers['content-type']
+    && request.headers['content-type'].match('json\\+fhir')) {
```
We should remove this matching on `json+fhir`. This is a very old remnant from when CDS Hooks was using FHIR model for its API.

kpshek commented on index.js:
```diff
@@ -0,0 +1,73 @@
+const express = require('express');
```
Both the Discovery and CDS Service endpoints return a `Content-Type` of `text/html`. We should be returning `application/json; charset=utf-8`.

kpshek commented on README.md:
```diff
@@ -1,2 +1,20 @@
 # Sandbox CDS Services
```
Add a **License** section to the bottom of this README.md:

```
   Copyright 2017 Cerner Innovation, Inc

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
```

Also, create a `NOTICE` file with the exact same contents as the above text.

kpshek replied:
Add a **Testing and Linting** section to this README.md.

zplata commented:
@kpshek Updated branch with changes requested. In addition:
- Updated the app structure for modularity, using Express Router
- Added tests using the Jest library (updated README)

mjhenkes commented on index.js:
```diff
@@ -0,0 +1,45 @@
+
+
+app.use('/cds-services', cdsServices);
+
```
You should probably add code to catch unhanded routes and errors. Something like this.
```  // catch 404 and forward to error handler
  app.use((req, res, next) => {
    const err = new Error('Not Found');
    err.status = 404;
    next(err);
  });

  // error handler
  app.use((err, req, res) => {
    // set locals, only providing error in development
    res.locals.message = err.message;
    res.locals.error = req.app.get('env') === 'development' ? err : {};

    // render the error page
    res.status(err.status || 500);
    res.render('error');
  });
```

kpshek replied:
👍

zplata replied:
Addressed in https://github.com/cds-hooks/sandbox-cds-services/pull/1/commits/5611a7fbba491d40267a41e095c0585c566fc7da

mjhenkes commented on services/patient-greeting.js:
```diff
@@ -0,0 +1,37 @@
+router.use((request, response, next) => {
+  const data = request.body;
+  if (!(data && data.prefetch && data.prefetch.patient && data.prefetch.patient.resource)) {
+    response.sendStatus(200);
```
If there is no prefetch patient resource you return a 200?
Should this be a 400 (bad request)?
I think normally in this case the cds service would ideally fall back to retrieving the patient information itself, but since that isn't implemented and the expected request doesn't contain the information provided, this seems like a bad request.

zplata replied:
@mjhenkes Just checked the [latest spec](http://cds-hooks.org/#a-performance-tweak), and it looks like on failure to receive necessary data from prefetch, services should respond with a `HTTP 412 Precondition Failed status code`. I can make that change here. In a future enhancement we can add that logic to retrieve the patient information itself, if allowable. Thoughts?

kpshek replied:
I want the CDS Service to retrieve Patient resource itself rather than returning a 412. The 412 status should be reserved for exceptional cases where the CDS Service is not capable of calling the FHIR server on its own.

mjhenkes replied:
Do we need to go that far for this demo service? In a real cds services absolutely, but it might be overkill here.

kpshek replied:
Given that the service we're replacing doesn't do this I'm fine if we log an issue and make that change later. I'm just sensitive to people looking at the code for this service and using the practices here in their real services.

The expectation of prefetching being optional has been a continual challenge for us because the Sandbox honors all prefetch requests.

mjhenkes replied:
Cool, lets log an issue and make this change in a separate pull request.

zplata replied:
Issue logged: https://github.com/cds-hooks/sandbox-cds-services/issues/3

kpshek commented on discovery/cds-services.js:
```diff
@@ -0,0 +1,18 @@
+  const discoveryEndpointServices = {
+    services: serviceDefinitions,
+  };
+  response.send(JSON.stringify(discoveryEndpointServices, null, 2));
```
I think we should be using [response.json](http://expressjs.com/en/api.html#res.json) instead.

zplata replied:
Addressed in https://github.com/cds-hooks/sandbox-cds-services/pull/1/commits/5611a7fbba491d40267a41e095c0585c566fc7da

kpshek commented on services/patient-greeting.js:
```diff
@@ -0,0 +1,37 @@
+      indicator: 'info',
+    }],
+  };
+  response.send(JSON.stringify(card, null, 2));
```
I think we should be using [response.json](http://expressjs.com/en/api.html#res.json) instead.

mjhenkes replied:
the only difference is that what zach has is easier to read in browsers. It adds in the tabs like githubs apis.

kpshek replied:
This line in index.js will accomplish this:

```javascript
app.set('json spaces', '  ')
```

mjhenkes replied:
Ooh, nice 👍
